compiler/rustc_lint/src/context.rs

```diff
@@ -888,31 +890,39 @@ pub trait LintContext: Sized {
         span: S,
         decorator: impl for<'a> DecorateLint<'a, ()>,
     ) {
-        self.lookup(lint, Some(span), |diag| decorator.decorate_lint(diag));
+        self.lookup(lint, Some(span), decorator.msg(), |diag| decorator.decorate_lint(diag));
     }
 
     fn struct_span_lint<S: Into<MultiSpan>>(
         &self,
         lint: &'static Lint,
         span: S,
-        decorate: impl for<'a> FnOnce(LintDiagnosticBuilder<'a, ()>),
+        msg: impl Into<DiagnosticMessage>,
+        decorate: impl for<'a, 'b> FnOnce(
+            &'b mut DiagnosticBuilder<'a, ()>,
+        ) -> &'b mut DiagnosticBuilder<'a, ()>,
```

Comment:
It seems weird to take a mutable reference and then return it. Why did you pick this scheme instead of just taking an `&mut DiagnosticBuilder` argument? For a smaller diff you could also just add a `'b` lifetime to `LintDiagnosticBuilder` and put the diagnostic builder inside of it as a field.

Reply:
oh cjgillot already said that

Reply:
This is weird, but at the same time I see the rationale. It makes sense to me in context, given how these APIs end up being used.